components/core/src/env.rs

```diff
@@ -72,6 +72,75 @@ pub fn var_os<K: AsRef<OsStr>>(key: K) -> std::option::Option<OsString> {
     }
 }
 
+#[macro_export]
+macro_rules! env_config {
+    (
+        $wrapping_type:ident,
+        $wrapped_type:ty,
+        $env_var:ident,
+        $default_value:expr,
+        $from_str_err_type:ty,
+        $from_str_input:ident,
+        $from_str_return:expr
+    ) => {
+        use $crate::env::Config as _;
+
+        struct $wrapping_type($wrapped_type);
+        // A little trickery to avoid env var name collisions:
+        // This enum can't ever be instantiated, but the compiler will give
+        // an error if two invocations in a namespace give the same env_var.
+        // It'd be nice to make this work globally, but I don't see a good way
+        #[allow(non_camel_case_types, dead_code)]
+        enum $env_var {}
```

Comment:
Here's what it generates if I made the same mistake I made in https://github.com/habitat-sh/habitat/pull/6565/commits/2bd5ad6ca5c5b25553e894d0322bda23b43c1fc4:
```
error[E0428]: the name `HAB_THREAD_ALIVE_THRESHOLD_SECS` is defined multiple times
   --> components/common/src/lib.rs:162:5
    |
162 | /     habitat_core::env_config_duration!(ThreadAliveThreshold,
163 | |                                        HAB_THREAD_ALIVE_THRESHOLD_SECS,
164 | |                                        Duration::from_secs(5 * 60));
    | |____________________________________________________________________^ `HAB_THREAD_ALIVE_THRESHOLD_SECS` redefined here
165 | /     habitat_core::env_config_duration!(ThreadAliveCheckDelay,
166 | |                                        HAB_THREAD_ALIVE_THRESHOLD_SECS,
167 | |                                        Duration::from_secs(60));
    | |________________________________________________________________- previous definition of the type `HAB_THREAD_ALIVE_THRESHOLD_SECS` here
    |
    = note: `HAB_THREAD_ALIVE_THRESHOLD_SECS` must be defined only once in the type namespace of this module
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

Reply:
This is _really_ slick!
![tenor-73439784](https://user-images.githubusercontent.com/207178/57944408-e09f0b80-78a4-11e9-96c2-9d87b6a080b1.gif)